miner/scroll_worker.go

```diff
@@ -261,44 +292,85 @@ func (w *worker) close() {
 	w.wg.Wait()
 }
 
+// checkHeadRowConsumption will start some initial workers to CCC check block close to the HEAD
+func (w *worker) checkHeadRowConsumption() error {
+	checkStart := uint64(1)
+	numOfBlocksToCheck := uint64(w.config.CCCMaxWorkers + 1)
+	currentHeight := w.chain.CurrentHeader().Number.Uint64()
+	if currentHeight > numOfBlocksToCheck {
+		checkStart = currentHeight - numOfBlocksToCheck
+	}
+
+	for curBlockNum := checkStart; curBlockNum <= currentHeight; curBlockNum++ {
+		block := w.chain.GetBlockByNumber(curBlockNum)
+		// only spawn CCC checkers for blocks with no row consumption data stored in DB
+		if rawdb.ReadBlockRowConsumption(w.chain.Database(), block.Hash()) == nil {
+			if err := w.asyncChecker.Check(block); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
 // mainLoop is a standalone goroutine to regenerate the sealing task based on the received event.
 func (w *worker) mainLoop() {
 	defer w.wg.Done()
+	defer w.asyncChecker.Wait()
 	defer w.txsSub.Unsubscribe()
 	defer w.chainHeadSub.Unsubscribe()
 
-	deadCh := make(chan *pipeline.Result)
-	pipelineResultCh := func() <-chan *pipeline.Result {
-		if w.currentPipeline == nil {
-			return deadCh
+	var err error
+	for {
+		if _, isRetryable := err.(retryableCommitError); isRetryable {
+			if _, err = w.tryCommitNewWork(time.Now(), w.current.header.ParentHash, w.current.reorgReason); err != nil {
+				continue
+			}
+		} else if err != nil {
+			log.Error("failed to mine block", "err", err)
+			w.current = nil
+		}
+
+		// check for reorgs first to lower the chances of trying to handle another
+		// event eventhough a reorg is pending (due to Go `select` pseudo-randomly picking a case
+		// to execute if multiple of them are ready)
+		select {
+		case trigger := <-w.reorgCh:
+			err = w.handleReorg(&trigger)
+			continue
+		default:
 		}
-		return w.currentPipeline.ResultCh
-	}
 
-	for {
 		select {
 		case <-w.startCh:
-			w.startNewPipeline(time.Now().Unix())
-		case <-w.chainHeadCh:
-			w.startNewPipeline(time.Now().Unix())
-		case result := <-pipelineResultCh():
-			w.handlePipelineResult(result)
+			if err := w.checkHeadRowConsumption(); err != nil {
+				log.Error("failed to start head checkers", "err", err)
+				return
+			}
+
+			_, err = w.tryCommitNewWork(time.Now(), w.chain.CurrentHeader().Hash(), nil)
+		case trigger := <-w.reorgCh:
+			err = w.handleReorg(&trigger)
+		case chainHead := <-w.chainHeadCh:
+			if w.isCanonical(chainHead.Block.Header()) {
```

Comment:
Will this only return false if we had previously reorged this block?

Reply:
Correct.